### PR TITLE
Added unit tests to improve code coverage

### PR DESCRIPTION
**Pull request by nbhoski**

### CODE REVIEW

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java [hunk 1]
```diff
@@ -279,4 +282,134 @@ public void verifyCustomCommandInvokedForBatchMode() throws Exception {
...
+        Assert.assertFalse(runOption.getBooleanByName("fakeCommand"));
+    }
+    
+    /*
+     * Test to verify when MATLAB version is older the R2017a
+     */
+
+    @Test
+    public void verifyMatlabVersionOlderThanR17a() throws Exception {
+        this.matlabBuilder.setLocalMatlab(getMatlabroot("R2018b"));
```
Comment (jwilkick-mw) [inline src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: Should the version string in here be for 17a or an older release?  I'm just going by the name of this test case and I'm not sure if I'm interpreting this correctly.

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java [hunk 2]
```diff
@@ -279,4 +282,134 @@ public void verifyCustomCommandInvokedForBatchMode() throws Exception {
...
+        RunTestsAutomaticallyOption runOption = new RunTestsAutomaticallyOption();
+        runOption.setTaCoberturaChkBx(true);
+        runOption.setTaJunitChkBx(true);
+        runOption.setTatapChkBx(true);
+        this.matlabBuilder.setTestRunTypeList(runOption);
+        project.getBuildersList().add(this.matlabBuilder);
+        FreeStyleBuild build = project.scheduleBuild2(0).get();
+        jenkins.assertLogContains("-batch", build);
+        jenkins.assertLogContains("true,true,true", build);
```
Comment (jwilkick-mw) [inline src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: Is there any chance this test could pick up the wrong output?  Is there some way to make it clearer that these boolean values came from the checkboxes?

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java [hunk 3]
```diff
@@ -279,4 +282,134 @@ public void verifyCustomCommandInvokedForBatchMode() throws Exception {
...
+    
+    /*
+     * Test to verify default value of getStringByName() when Automatic test mode. 
+     */
+
+    @Test
+    public void verifyDefaultValueOfgetStringByName() throws Exception {
+        this.matlabBuilder.setLocalMatlab(getMatlabroot("R2018b"));
+        RunTestsAutomaticallyOption runOption = new RunTestsAutomaticallyOption();    
+        Assert.assertNull(runOption.getStringByName("fakeChkBox"));
```
Comment (jwilkick-mw) [inline src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: It seems like this is testing the behavior of a different class. Is there a separate test for the RunTestsAutomaticallyOption that could do this?

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java [hunk 4]
```diff
@@ -279,4 +282,134 @@ public void verifyCustomCommandInvokedForBatchMode() throws Exception {
...
+
+    @Test
+    public void verifyMatlabVersionOlderThanR17a() throws Exception {
+        this.matlabBuilder.setLocalMatlab(getMatlabroot("R2018b"));
+        RunTestsAutomaticallyOption runOption = new RunTestsAutomaticallyOption();
+        runOption.setTaCoberturaChkBx(true);
+        runOption.setTaJunitChkBx(true);
+        runOption.setTatapChkBx(true);
+        this.matlabBuilder.setTestRunTypeList(runOption);
+        project.getBuildersList().add(this.matlabBuilder);
```
Comment (jwilkick-mw) [inline src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: Noticed that some of the code from this point to the end of the method is repeated in another test.  Could you pull this out into a different function?

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java [hunk 5]
```diff
@@ -279,4 +282,134 @@ public void verifyCustomCommandInvokedForBatchMode() throws Exception {
...
+        WebAssert.assertTextPresent(page, TestMessage.getValue("Builder.matlab.root.empty.error"));
+    }
+    
+    /*
+     * Test To verify UI does not throw any error when valid MATLAB root entered
+     * 
+     */
+
+    @Test
+    public void verifyInvalidValidMatlabRoot() throws Exception {
```
Comment (jwilkick-mw) [inline src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: Given the comment, could you name this:  verifyValidMatlabRootDoesntDisplayError?

Or, looking at the actual test code, it seems like it **is** checking that an invalid root shows an error. I'm a bit confused by this one.

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java [hunk 6]
```diff
@@ -279,4 +282,134 @@ public void verifyCustomCommandInvokedForBatchMode() throws Exception {
...
+        WebAssert.assertTextPresent(page, TestMessage.getValue("Builder.matlab.cobertura.support.warning"));
+    }
+    
+    /*
+     * Test To verify UI displays Cobertura Error message when invalid MATLAB root entered.
+     * 
+     */
+
+    @Test
+    public void verifyCoberturaError() throws Exception {
```
Comment (jwilkick-mw) [inline src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: Should this be testing that Builder.matlab.cobertura.support.warning is there?  Otherwise it seems like the invalid matlab root test might cover this already.

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java (hunk 1 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: Thanks for pointing out this has to be older version. fixed it by using R2016b version. it required some change in the getMatlabroot() method as well.

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java (hunk 2 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: these have chance of taking default values if someone changes the .jelly file with corresponding web element names , but forgets to add it in action class. i don't see direct way to map these to confirm if its coming from UI other than writing some Web automation (functional tests) . I got your point though.

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java (hunk 3 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: No , we do not have separate test for this class as this is part of MatlabBuilder class by design. to support multi dropdown.options . this test actually is targeting more of line coverage / line coverage.

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java (hunk 4 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: Done.

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java (hunk 5 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: Good catch , its copy paste error, This test case checks if error is displayed if invalid MATLAB root is entered. i have changed the name to verifyInvalidMatlabRootDisplaysError and verifyValidMatlabRootDoesntDisplayError

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java (hunk 6 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: Correct , this Error is already verified however both these scenario uses MatlabReleaseInfo to check the MATLAb version and both of them they throw version not found exception. hence in the exception case our code coverage was not hitting the exception block. to cover the same this test is added.

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java [hunk 7]
```diff
@@ -87,9 +88,17 @@ public void testTearDown() {
 
     private String getMatlabroot(String version) throws URISyntaxException {
         ClassLoader classLoader = MatlabBuilderTest.class.getClassLoader();
-        String matlabRoot = new File(
-                classLoader.getResource("versioninfo/" + version + "/VersionInfo.xml").toURI())
-                        .getAbsolutePath().replace(FileSeperator + "VersionInfo.xml", "");
+        String matlabRoot;
```
Comment (jwilkick-mw) [inline src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: This bit of code is kind of hard to follow.  Could you assign the result of `classLoader.getResource(...)` to a variable and then use that in the if condition?

```java
String matlabRoot;
URL versionInfo = classLoader.getResource("versioninfo/"+version+"/VersionInfo.xml");

if(versionInfo!=null){
    matlabRoot = new File(versionInfo.toURI()).getAbsolutePath();
}else{
    matlabRoot = new File(classLoader.getResource("versioninfo/R2017a/VersionInfo.xml").toURI()).getAbsolutePath().replace("R2017a",version);
}

return matlabRoot.replace(FileSeparator+"VersionInfo.xml","");
```
Not everyone is a fan of Optionals, but I could also see the code like this:

```java
private String getMatlabroot(String version) throws URISyntaxException {
     String defaultVersionInfo = "versioninfo/R2017a/VersionInfo.xml";
     String userVersionInfo = "versioninfo/"+version+"/VersionInfo.xml";

      Optional<URL> matlabRootURL = Optional.ofNullable(getResource(userVersionInfo)).orElseGet(() -> getResource(defaultVersionInfo));
      
      File matlabRoot = new File(matlabRootURL.get().toURI());

      return matlabRoot.getAbsolutePath().replace(FileSeperator+"VersionInfo.xml","").replace("R2017a",version);      
}

private URL getResource(String resource){
    return MatlabBuilderTest.class.getClassLoader().getResource(resource);   
}
```

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java [hunk 8]
```diff
@@ -412,4 +409,19 @@ public void verifyCoberturaError() throws Exception {
         String filteredPageText = pageText.replaceFirst(TestMessage.getValue("Builder.invalid.matlab.root.error"), "");
         Assert.assertTrue(filteredPageText.contains(TestMessage.getValue("Builder.invalid.matlab.root.error")));
     }
+    
+    /*
+     * Private helper methods for tests
+     */
+    
+    private FreeStyleBuild getBuildforRunTestAutomatically() throws InterruptedException, ExecutionException {
```
Comment (jwilkick-mw) [inline src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: The helper method is a good addition.

File: src/test/java/com/mathworks/ci/MatlabBuilderTest.java (hunk 7 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/MatlabBuilderTest.java]: Thanks @jwilkick-mw  . it was a good suggestion for using Optional <T> class
